edgedb-tokio/src/tutorial.md

```diff
@@ -0,0 +1,475 @@
+}
+```
+
+Most variants of the `Value` enum correspond to a Rust type from the standard library, while some are from the `edgedb-protocol` crate and will have to be constructed. For example, this query expecting a `bigint` will return an error as it receives a `20` (an `i32`) but EdgeDB is expecting a `bigint`:
+
+```rust
+let query = "select <bigint>$0";
+let argument = 20;
+let query_res: Result<Value, _> = client.query_required_single(query, &(argument,)).await;
+assert!(format!("{query_res:?}").contains("expected std::int32"));
+```
+
+Instead, first construct a `BigInt` from the `i32` and pass that in as an argument:
+
+```rust
+use edgedb_protocol::model::BigInt;
+
+let query = "select <bigint>$0";
+let bigint_arg = BigInt::from(20);
+let query_res: Result<Value, _> = client.query_required_single(query, &(bigint_arg,)).await;
+assert_eq!(
+    format!("{query_res:?}"),
+    "Ok(BigInt(BigInt { negative: false, weight: 0, digits: [20] }))"
+);
+```
+
+## Casting inside the EdgeDB compiler
+
+EdgeDB requires arguments to have a cast in the same way that Rust requires a type declaration in function signatures. As such, arguments in queries are used as type specification for the EdgeDB compiler, not to cast from queries from the Rust side. Take this query as an example:
+
+```rust
+let query = "select <int32>$0";
+```
+
+This simply means "select an argument that must be an `int32`", not "take the received argument and cast it into an `int32`".
+
+As such, this will return an error:
+
+```rust
+let query = "select <int32>$0";
+let argument = 9i16; // Rust client will expect an int16
+let query_res: Result<Value, _> = client.query_required_single(query, &(argument,)).await;
+assert!(query_res
+    .unwrap_err()
+    .to_string()
+    .contains("expected std::int16"));
+```
+
+## Using JSON
+
+EdgeDB can cast any type to JSON with `<json>`, but the `_json` methods don't require this cast in the query. This result can be turned into a String and used to respond to some JSON API request directly, unpacked into a struct using `serde` and `serde_json`, etc.
+
+```rust
+#[derive(Debug, Deserialize)]
+pub struct Account {
+    pub username: String,
+    pub id: Uuid,
+}
+
+// No need for <json> cast here
+let query = "select Account { 
+      username,
+      id
+    } filter .username = <str>$0;";
+
+// Assuming we know there will only be one result we can use query_single_json;
+// otherwise query_json which returns a map of json
+let json_res = client
+    .query_single_json(query, &("SomeUserName",))
+    .await?
+    .unwrap();
+
+// Format: {"username" : "SomeUser1", "id" : "7093944a-fd3a-11ed-a013-c7de12ffe7a9"}
+let as_string = json_res.to_string();
+let as_account: Account = serde_json::from_str(&json_res)?;
+```
+
+## Execute
+
+The `execute` method was added in June 2023 and will show up inside the next version of the `edgedb-tokio` crate. In the meantime it can be used by setting your edgedb dependencies to the git repo:
```

Comment:
> …inside the next version of the`edgedb-tokio` crate

"Inside" instead of "in" because it's a "crate?"

Reply:
```suggestion
The `execute` method was added in June 2023 and will show up inside the next version of the `edgedb-tokio` crate. In the meantime it can be used by setting your EdgeDB dependencies to the git repo:
```